meetings/working-groups/extensions/anonymous-extension-declarations.md

```diff
@@ -0,0 +1,368 @@
+public static class NullableExtensions
+{
+    extension(string?)
+    {
+        public string AsNotNull => this is null ? "" : this;
+        [param:NotNullWhen(false)] public bool IsNullOrEmpty => this is null or [];
+    }
+    extension(ref string?)
+    {
+        [param:NotNull] public void MakeNotNull() => this ??= "";
+    }
+    extension<T> (T) where T : class?
+    {
+        [param:NotNull] public void ThrowIfNull() => ArgumentNullException.ThrowIfNull(this);
+    }
+}
+```
+
+### Lowering
+
+Extension declarations give rise to static classes nested within the enclosing static class. These inner static classes are generated with unspeakable names and do not pollute the namespace of the enclosing static class. 
+
+Within a given enclosing static class, extension declarations with the same number of type parameters and equivalent constraints are merged into a single static class even if they differ on underlying type, ref-kind or type parameter names. Thus the minimum number of nested classes are generated to represent the combinations of type parameters and constraints present across all the extension declarations within the enclosing static class:
+
+``` c#
+public static class MyExtensions
+{
+    extension<TSource>(Span<TSource>) where TSource : class?
+    {
+        ...
+    }
+    extension<TElement>(Span<TElement?>) where TElement : class
+    {
+        ...
+    }
+}
+```
+
+Because the signatures are sufficiently equivalent, this is lowered to just one class:
+
+``` c#
+public static class MyExtensions
+{
+    public static class __E1<__T1> where __T1 : class
+    {
+        ...
+    }
+}
+```
+
+The requirement is not unlike that between the two parts of a partial method in today's C#.
+
+The members themselves are generated as static members. For instance members, insofar as it is possible to represent in IL, the same kind of member is generated, but with an extra first parameter for the underlying value. Where not possible, the body is represented as one or two (for accessors) static methods, and attributes are emitted to record their original member kind.
+
+The generated first parameter includes any ref kind specified in the extension declaration, and any attributes on the member which have the `param` target specifier:
+
+``` c#
+public static class NullableExtensions
+{
+    extension<T> (T) where T : class?
+    {
+        [param:NotNull] public void ThrowIfNull() => ArgumentNullException.ThrowIfNull(this);
+    }
+}
+```
+
+Generates:
+
+``` c#
+public static class NullableExtensions
+{
+    public static class __E1<__T1> where T1__ : class
+    {
+        public static void ThrowIfNull([NotNull] __this) => ArgumentNullException.ThrowIfNull(__this);
+    }
+}
+```
+
+### Checking
+
+__Inferrability:__ The underlying type of an extension declaration must make use of all the type parameters of that extension declaration, so that it is always possible to infer the type arguments when applied to the underlying type.
```

Comment:
Even with this requirement, it will not always be possible to infer the type arguments. E.g. `extension<T>(IAnyInterface<T>)`, when you dot off a receiver that implements `IAnyInterface<string>` and `IAnyInterface<Utf8String>` simultaneously.

Reply:
As a potential solution to this, we could use this syntax for disambiguation:

```cs
Enumerable.extension<Xyz>(receiver).Member...
```

If we have something like this, then we don't need the rule that the type parameter appears in the extension's underlying type.

Reply:
Casting the receiver should always be able to resolve this, so never mind!